[KEP-4006] Blog Post: Streaming Transition from SPDY to WebSockets is Beta

seans3 opened the pull request:
### Streaming Transition from SPDY to Websocket Blog Post

https://github.com/kubernetes/enhancements/issues/4006

a-mccarthy commented:
/hold

Pending assignment of publication date (by release comms)

We'll also keep this held until the v1.31 release has happened

edithturn commented:
Hey @seans3, it's Edith Puclla from the 1.31 Release comms team.

A friendly reminder that the deadline for the blog to be ready for review is **July 26, 2024.**

Let me know if there's anything we can do to help with content, review, or anything else! 🙂

seans3 commented:
> Hello Sean Sullivan 🙂 , it's Edith Puclla from the 1.31 Release comms team.
> 
> A friendly reminder that the deadline for the blog to be ready for review is **July 26, 2024.**
> 
> Let me know if there's anything we can do to help with content, review, or anything else! 🙂

Acknowledged--thanks :) I may be in touch to get more clarity on expectations.

seans3 commented:
> Hello @seans3 🙂 , it's Edith Puclla from the 1.31 Release comms team.
> 
> A friendly reminder that the deadline for the blog to be ready for review is **July 26, 2024.**
> 
> Let me know if there's anything we can do to help with content, review, or anything else! 🙂

Please have a look and let me know what you think. And if there is a more solid date, I'll update it in the PR. Thanks

tengqm commented:
/approve

seans3 commented:
/assign @edithturn

seans3 commented:
/assign @liggitt 
/assign @aojea

aojea commented on content/en/blog/_posts/2024-XX-XX-websocket-transition.md:
```diff
@@ -0,0 +1,132 @@
+---
+layout: blog
+title: 'Streaming Transitions from SPDY to WebSockets'
+date: 2024-XX-XX
```
is this on purpose?

seans3 replied:
Updated the date, now that it's been selected.

aojea commented on content/en/blog/_posts/2024-XX-XX-websocket-transition.md:
```diff
@@ -0,0 +1,132 @@
+## Why have we modified the streaming protocol ?
+
+Consider the following situation: you’ve configured a proxy or gateway in front
+of your Kubernetes cluster, but you’ve discovered four kubectl commands no
```
do we need to clarify that the proxy or gateway is in fron of the kubernetes cluster control-plane/apiserver?

seans3 replied:
This entire paragraph has been re-written. Please have a look.

aojea commented on content/en/blog/_posts/2024-XX-XX-websocket-transition.md:
```diff
@@ -0,0 +1,132 @@
+longer work: kubectl exec, kubectl cp, kubectl attach, and kubectl port-forward.
+For these kubectl commands, the HTTP REST requests are upgraded to streaming
+connections. But the streaming protocol used for these upgraded connections is
+SPDY/3.1, which has been **deprecated for eight years**. Your newly configured
+proxy or gateway does not support this out-of-date streaming protocol. As of the
+v1.31 release, however, by default all four of these kubectl commands now
+implement the more modern [Websocket streaming protocol](https://datatracker.ietf.org/doc/html/rfc6455).
```
I would focus more on standardization than in age, as SPDY was never standardized and websockets is widely used standard

aojea commented on content/en/blog/_posts/2024-XX-XX-websocket-transition.md:
```diff
@@ -0,0 +1,132 @@
+longer work: kubectl exec, kubectl cp, kubectl attach, and kubectl port-forward.
+For these kubectl commands, the HTTP REST requests are upgraded to streaming
+connections. But the streaming protocol used for these upgraded connections is
+SPDY/3.1, which has been **deprecated for eight years**. Your newly configured
```
I would also mention that was never standardized

seans3 replied:
Done.

aojea commented on content/en/blog/_posts/2024-XX-XX-websocket-transition.md:
```diff
@@ -0,0 +1,132 @@
+  [Sean Sullivan](https://github.com/seans3) (Google)
+---
+
+With Kubernetes 1.31, PortForward over Websockets is moving to beta.
+RemoteCommand over Websockets previously graduated to beta in v1.30.
```
websockets were already supported (modulo known bugs you fixed) but I think the feature is about making it the default in kubectl, right?

seans3 replied:
I have completely re-written this initial sentence, and I think it addresses your feedback and simplifies the sentence. Please let me know what you think.

edithturn commented:
Hi @seans3! Just a reminder that the publish date for this blog is scheduled for August 20th 🙂

edithturn commented:
> /assign @edithturn

Hello @seans3 
This is just a friendly reminder that the review date is today. When you have a moment, could you please give us an update? 
I'd love to mark it as "ready for review." Let me know if there's anything I can assist you with 🙂

Thanks a bunch!

Edith Puclla
Comms Shadow

seans3 commented:
> > /assign @edithturn
> 
> Hello @seans3 This is just a friendly reminder that the review date is today. When you have a moment, could you please give us an update? I'd love to mark it as "ready for review." Let me know if there's anything I can assist you with 🙂
> 
> Thanks a bunch!
> 
> Edith Puclla Comms Shadow

Ready for review.

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,106 @@
+a shell in a running container from your local workstation and run commands in
+the container.
+
+## Why have we modified the streaming protocol ?
```
```suggestion
## Why the streaming protocol is changing
```
nit

seans3 replied:
Fixed.

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,106 @@
+Currently, Kubernetes uses the SPDY/3.1 protocol for upgraded streaming
+connections. SPDY/3.1 has been deprecated for eight years. Many modern proxies,
+gateways, and load balancers no longer support the protocol. As a result, you
+might notice that commands like kubectl cp, kubectl attach, kubectl exec, and
+kubectl port-forward stop working when you try to access your cluster through a
```
```suggestion
might notice that commands like `kubectl cp`, `kubectl attach`, `kubectl exec`, and
`kubectl port-forward` stop working when you try to access your cluster through a
```

seans3 replied:
Fixed.

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,106 @@
+programming languages. The Websocket protocol is more widely supported by modern
+proxies and gateways than SPDY.
+
+## How do these streaming API’s work ?
```
```suggestion
## How streaming APIs work
```

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,106 @@
+
+## How do these streaming API’s work ?
+
+Streaming API’s work by upgrading an HTTP connection to a streaming connection.
+An HTTP upgrade request for running the date command on an nginx container
+within a cluster could look like:
```
```suggestion
Kubernetes upgrades HTTP connections to streaming connections by adding
specific upgrade headers to the originating HTTP request. For example, an HTTP
upgrade request for running the `date` command on an `nginx` container within
a cluster is similar to the following:
```

seans3 replied:
Fixed.

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
````diff
@@ -0,0 +1,106 @@
+    User-Agent: kubectl/v1.30.1 (linux/amd64) kubernetes/6911225
+```
+
+If the container runtime supports the Websocket streaming protocol and at least
````
```suggestion
If the container runtime supports the WebSocket streaming protocol and at least
```

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,106 @@
+and process exit code data) for this shell interaction is then streamed over this
+upgraded connection.
+
+## How do I leverage the new Websocket streaming protocol ?
```
```suggestion
## How to use the new Websocket streaming protocol
```

seans3 replied:
Fixed.

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,106 @@
+
+## How do I leverage the new Websocket streaming protocol ?
+
+If your cluster version and kubectl are greater than or equal to v1.29, there
+are two control plane feature flags and two kubectl feature flags that govern
+the use of the new Websocket streaming protocol. As of v1.31 release, all of the
+following feature flags are in beta and are enabled by default.
```
```suggestion
If your cluster and kubectl are on version 1.29 or later, there are two
control plane feature flags and two kubectl feature flags that govern
the use of the new WebSocket streaming protocol. In version 1.31 and
later, all of the following feature flags are in beta and are enabled by
default:
```

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,106 @@
+  - `KUBECTL_PORT_FORWARD_WEBSOCKETS`
+      - `kubectl port-forward`
+
+## Where can I find more information on streaming API’s ?
```
```suggestion
## Learn more about streaming APIs
```

seans3 replied:
Fixed.

shannonxtreme reviewed (commented):
I can't LGTM because I'm an author but here are some comments, overall feels clean and informative.

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+stop working when you try to access your cluster through a proxy or gateway.
+
+As of Kubernetes 1.31, we're modifying the streaming protocol that Kubernetes
+uses for these commands to the more modern [Websocket streaming protocol](https://datatracker.ietf.org/doc/html/rfc6455).
```
@seans3 pls replace Websocket with WebSocket throughout the file :)

seans3 replied:
Fixed.

shannonxtreme commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+  [Shannon Kularathna](https://github.com/shannonxtreme) (Google)
+---
+
+In Kubernetes 1.31, by default kubectl now uses the WebSocket protocol
+instead of SPDY for streaming.
```
```suggestion
In Kubernetes 1.31 and later, Kubernetes uses the WebSocket protocol instead of
SPDY when upgrading HTTP connections from `kubectl` to streaming connections.
```

shannonxtreme replied:
"connections from kubectl" might be inaccurate

sftim replied:
For me (and I'm an editor for the blog), the original wording is OK; this isn't documentation so being more conversational is OK.

shannonxtreme reviewed (commented):
This looks fine, left three comments for quality. The intro sentence could be clearer, commented for that

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+
+## Why the streaming protocol is changing
+
+Currently, Kubernetes uses the SPDY/3.1 protocol for upgraded streaming
```
This will be published _after_ the release has happened, so: (eg)
```suggestion
Before the v1.31 release, Kubernetes used the SPDY/3.1 protocol by default when upgrading
existing connections to streaming
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+`kubectl cp`, `kubectl attach`, `kubectl exec`, and `kubectl port-forward`
+stop working when you try to access your cluster through a proxy or gateway.
+
+As of Kubernetes 1.31, we're modifying the streaming protocol that Kubernetes
```
```suggestion
As of Kubernetes 1.31, we've modified the streaming protocol that a Kubernetes client
(such as `kubectl`)
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
````diff
@@ -0,0 +1,107 @@
+upgrade request for running the `date` command on an `nginx` container within
+a cluster is similar to the following:
+
+```shell
````
Don't mark this as shell as it's not a shell script. Try `console` instead.

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
````diff
@@ -0,0 +1,107 @@
+with a successful `101 Switching Protocols` status, along with the negotiated
+subprotocol version:
+
+```shell
````
Don't mark this as shell as it's not a shell script. Try console instead.

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+later, all of the following feature flags are in beta and are enabled by
+default:
+
+- Control Plane Feature Flags
```
```suggestion
- [Feature gates](/docs/reference/command-line-tools-reference/feature-gates/)
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+default:
+
+- Control Plane Feature Flags
+  - TranslateStreamCloseWebsocketRequests
+      - `.../exec`
+      - `.../attach`
+  - PortForwardWebsockets
+      - `.../port-forward`
```
```suggestion
  - `TranslateStreamCloseWebsocketRequests`
      - `.../exec`
      - `.../attach`
  - `PortForwardWebsockets`
      - `.../port-forward`
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+
+## How to use the new WebSocket streaming protocol
+
+If your cluster and kubectl are on version 1.29 or later, there are two
+control plane feature flags and two kubectl feature flags that govern
+the use of the new WebSocket streaming protocol. In version 1.31 and
+later, all of the following feature flags are in beta and are enabled by
+default:
```
```suggestion
If your cluster and kubectl are on version 1.29 or later, there are two
control plane feature gates and two kubectl environment variables that
govern the use of the WebSockets rather than SPDR. In Kubernetes 1.31,
all of the following feature gates are in beta and are enabled by
default:
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+      - `.../attach`
+  - PortForwardWebsockets
+      - `.../port-forward`
+- kubectl Feature Flags (environment variables)
```
```suggestion
- kubectl feature control environment variables
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+
+## Learn more about streaming APIs
+
+- [KEP - Transitioning from SPDY to Websockets](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets)
```
```suggestion
- [KEP 4006 - Transitioning from SPDY to Websockets](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets)
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+a shell in a running container from your local workstation and run commands in
+the container.
+
+## Why the streaming protocol is changing
```
```suggestion
## Why change the streaming protocol?
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,107 @@
+      - `kubectl attach`
+  - `KUBECTL_PORT_FORWARD_WEBSOCKETS`
+      - `kubectl port-forward`
+
```
(optionally)
```suggestion

If you're connecting to an older cluster but can manage the feature gate settings, turn on
`TranslateStreamCloseWebsocketRequests` and `PortForwardWebsockets` to try this new
behavior (you'll need Kubernetes v1.29 or v1.30 to try it out).

```

seans3 replied:
I added this small paragraph, but there is a subtle distinction I added. The `TranslateStreamCloseWebsocketRequests` feature gate is the only one available in v1.29. The `PortForwardWebsockets` feature gate is not available until v1.30. Please have a look at what I wrote, and let me know what you think.

sftim reviewed (commented):
Thanks

This is written as if the v1.31 release is in the future, which is true now but won't be at publication time. It's not ready yet, but the fixes needed are minor.

/approve cancel

seans3 commented:
> Thanks
> 
> This is written as if the v1.31 release is in the future, which is true now but won't be at publication time. It's not ready yet, but the fixes needed are minor.
> 
> /approve cancel

Thanks for the feedback. I think I've addressed the issues you raised. Please have a look.

edithturn commented:
> This is written as if the v1.31 release is in the future, which is true now but won't be at publication time. It's not ready yet, but the fixes needed are minor.

Hello @sftim , could you please 

> > Thanks
> > This is written as if the v1.31 release is in the future, which is true now but won't be at publication time. It's not ready yet, but the fixes needed are minor.
> > /approve cancel
> 
> Thanks for the feedback. I think I've addressed the issues you raised. Please have a look.

Hi @sftim! 
Could you please help us to check what @seans3 mentioned? 🙂

sftim commented:
> Hi @sftim! Could you please help us to check what @seans3 mentioned? 🙂

Any blog reviewer is welcome to move this forward, not just me.

mbianchidev commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,113 @@
+In Kubernetes, specific endpoints that are exposed as an HTTP or RESTful
+interface are upgraded to streaming connections, which require a streaming
+protocol. Unlike HTTP, which is a request-response protocol, a streaming
+protocol provides a persistent connection that's bi-directional, low-latency,
```
(nit) 

I would not contract and go for "that is" in full form

sftim replied:
For blog articles, we typically let the author write as them. No need to change here @seans3 if you are happy not to.

mbianchidev replied:
@sftim thanks for letting me know!

mbianchidev commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,113 @@
+`kubectl cp`, `kubectl attach`, `kubectl exec`, and `kubectl port-forward`
+stop working when you try to access your cluster through a proxy or gateway.
+
+As of Kubernetes v1.31, we've modified the streaming protocol that a Kubernetes
```
another nit

Full form "we have"

sftim replied:
For blog articles, we typically let the author write as them. No need to change here @seans3 if you are happy not to.

sftim replied:
However it's good to qualify “we“; for example, it could mean _we, Google, have modified this protocol_ or _we, SIG API Machinery, have&hellip;_ or even _we, Sean and Shannon, made the code changes&hellip;_

seans3 replied:
Fixed.

mbianchidev commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,113 @@
+      - `kubectl port-forward`
+
+If you're connecting to an older cluster but can manage the feature gate
+settings, turn on `TranslateStreamCloseWebsocketRequests` (available in
```
(nit)

Turn on X and Y -> Turn on both X and Y

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
````diff
@@ -0,0 +1,113 @@
+```
+
+At this point the TCP connection used for the HTTP protocol has changed to a
+streaming connection. Subsequent STDIN, STDOUT, and STDERR data (as well as TTY
````
```suggestion
streaming connection. Subsequent STDIN, STDOUT, and STDERR data (as well as terminal
```
What kind of data is it? If Windows containers can have a terminal, “TTY“ isn't quite right (Windows doesn't really use the term _teletype_, which TTY is an abbreviation of).

seans3 replied:
The TTY data is terminal window resizing data. So I've just termed it `terminal resizing data`. Please let me know what you think.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,113 @@
+    Connection: Upgrade
+    Upgrade: websocket
+    Sec-Websocket-Protocol: v5.channel.k8s.io
+    User-Agent: kubectl/v1.30.1 (linux/amd64) kubernetes/6911225
```
For the look of the thing, maybe this could be `kubectl 1.31.0` and an appropriate build number?

(we can do a fixup after the release, between the merge and actual publication).

seans3 replied:
Fixed by changing `v1.30.1` to `v1.31.0`.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,113 @@
+      - `kubectl port-forward`
+
+If you're connecting to an older cluster but can manage the feature gate
+settings, turn on `TranslateStreamCloseWebsocketRequests` (available in
```
(nit)
```suggestion
settings, turn on `TranslateStreamCloseWebsocketRequests` (added in
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,113 @@
+
+If you're connecting to an older cluster but can manage the feature gate
+settings, turn on `TranslateStreamCloseWebsocketRequests` (available in
+Kubernetes v1.29) and `PortForwardWebsockets` (available in Kubernetes
```
```suggestion
Kubernetes v1.29) and `PortForwardWebsockets` (added in Kubernetes
```

seans3 replied:
Fixed.

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,113 @@
+If you're connecting to an older cluster but can manage the feature gate
+settings, turn on `TranslateStreamCloseWebsocketRequests` (available in
+Kubernetes v1.29) and `PortForwardWebsockets` (available in Kubernetes
+v1.30) to try this new behavior.
```
For clarity:
```suggestion
v1.30) to try this new behavior. Version 1.31 of `kubectl` can automatically use the new behavior,
but you do need to connect to a cluster where the server-side features are explicitly enabled.
```

seans3 replied:
Added.

divya-mohan0209 reviewed (commented):
Thanks for this!
/lgtm

sftim commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,115 @@
+---
+layout: blog
+title: 'Streaming Transitions from SPDY to WebSockets'
```
Should the title mention Kubernetes 1.31? A typical post-release article does.

/lgtm cancel

edithturn replied:
It should be: 
`title: "Kubernetes 1.31: Streaming Transitions from SPDY to WebSockets"`

seans3 replied:
Fixed.

edithturn commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,115 @@
+`kubectl cp`, `kubectl attach`, `kubectl exec`, and `kubectl port-forward`
+stop working when you try to access your cluster through a proxy or gateway.
+
+As of Kubernetes v1.31, SIG API Machinery have modified the streaming
```
A minor typo:
It should be: "SIG API Machinery **has** modified the streaming..."

seans3 replied:
`have` changed to `has`

edithturn commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,115 @@
+
+## Learn more about streaming APIs
+
+- [KEP 4006 - Transitioning from SPDY to Websockets](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets)
```
nit:
It should be "WebSockets"

seans3 replied:
Fixed.

edithturn commented on content/en/blog/_posts/2024-08-20-websocket-transition.md:
```diff
@@ -0,0 +1,115 @@
+## Learn more about streaming APIs
+
+- [KEP 4006 - Transitioning from SPDY to Websockets](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4006-transition-spdy-to-websockets)
+- [RFC 6455 - The Websocket Protocol](https://datatracker.ietf.org/doc/html/rfc6455)
```
nit:
It should be "WebSockets"

seans3 replied:
Fixed.

divya-mohan0209 reviewed (commented):
Since all comments seem to have been addressed. Please feel free to unhold @mbianchidev!

/lgtm

mbianchidev commented:
/unhold

Kubernetes v1.31 has been released.

Thanks @divya-mohan0209 !
